File: app/strings.py
# -*- coding: utf-8 -*-
"""string manipulations"""

import logging
import unicodedata as ud
import hashlib
import string as str_lib

# Mapping for character replacements (umlauts, special characters)
REPLACEMENT_MAP = {
    # Cyrillic
    'Ё': 'Е', 'Й': 'И', 'Ъ': 'Ь',
    # German umlauts (uppercase)
    'Ä': 'AE', 'Ö': 'OE', 'Ü': 'UE', 'ß': 'SS',
    # Other common replacements (uppercase)
    'À': 'A', 'Á': 'A', 'Â': 'A', 'Ã': 'A', 'Å': 'A',
    'È': 'E', 'É': 'E', 'Ê': 'E', 'Ë': 'E',
    'Ì': 'I', 'Í': 'I', 'Î': 'I', 'Ï': 'I',
    'Ñ': 'N',
    'Ò': 'O', 'Ó': 'O', 'Ô': 'O', 'Õ': 'O',
    'Ù': 'U', 'Ú': 'U', 'Û': 'U',
    'Ý': 'Y',
    'Ç': 'C',
    'Þ': 'TH',
    'Đ': 'D',
    'Ł': 'L',
    'Ń': 'N',
    'Ǿ': 'O',
    'Ŕ': 'R',
    'Ś': 'S',
    'Ź': 'Z',
    'Ż': 'Z',
}


def strip_quotes(s: str) -> str:
    """
    Safe quote stripping: only removes quotes if they are symmetric
    (same type on both ends). Leaves internal/single-sided quotes alone.

    '"word word"' -> 'word word'
    '"word" word' -> '"word" word'  (no change, last char is not quote)
    '"word" "word"' -> '"word" "word"'  (internal quotes present)
    "'word'" -> 'word'
    """
    if s is None:
        return None
    s = s.strip()
    if len(s) >= 2:
        if (s[0] == '"' and s[-1] == '"') or (s[0] == "'" and s[-1] == "'"):
            return s[1:-1]
    return s


def unicode_upper(string: str) -> str:
    """custom UPPER + normalize for sqlite and other"""
    ret = ud.normalize('NFKD', string)
    ret = ret.upper()
    # Apply replacements from map
    for old, new in REPLACEMENT_MAP.items():
        ret = ret.replace(old, new)
    return ret


def strlist(string) -> str:
    """return string or first element of list"""
    if isinstance(string, str):
        return strnull(string)
    if isinstance(string, list):
        if string:
            return strnull(string[0])
        return strnull("")  # empty list
    return strnull(str(string))


def strnull(string) -> str:
    """return empty string if None, else return content"""
    if string is None:
        return ""
    return str(string)


def num2int(num: str, context: str) -> int:
    """number in string or something to integer"""
    try:
        ret = int(num)
        return ret
    # pylint: disable=W0703
    except Exception as ex:  # not exception, but error in data
        logging.error("Error: %s", str(ex))
        logging.error("Context: %s", context)
        return -1  # invalid num indicator, valid numbers >=0


def id2path(id: str):  # pylint: disable=W0622
    """1a2b3c.... to 1a/2b/1a2b3c..."""
    first = id[:2]
    second = id[2:4]
    return first + "/" + second + "/" + id


def id2pathonly(id: str):  # pylint: disable=W0622
    """1a2b3c.... to 1a/2b"""
    first = id[:2]
    second = id[2:4]
    return first + "/" + second


def string2filename(data: str) -> str:
    """remove fs-dangerous characters from data"""
    if not data:
        data = "-"
    data = data.replace('/', '⧸')
    data = data.replace('.', '_')
    return data


def make_id(name, name_as_is: bool = False) -> str:
    """get name, return md5 of normalized string"""
    name_str = "--- unknown ---"
    if name is not None and name != "":
        if isinstance(name, str):
            name_str = str(name)
        else:
            name_str = str(name, encoding='utf-8')
    norm_name = name_str if name_as_is else str_normalize(name_str)
    return hashlib.md5(norm_name.encode('utf-8')).hexdigest()


def str_normalize(string: str) -> str:
    """normalize string for make_id and compare"""
    if not string:
        return ""

    # Strip leading/trailing whitespace
    ret = string.strip()

    # Replace multiple spaces with single space
    while '  ' in ret:
        ret = ret.replace('  ', ' ')

    # Normalize all quote types to one canonical form (double quotes)
    # This ensures that 'name', "name", and «name» produce the same ID
    ret = ret.replace('«', '"')
    ret = ret.replace('»', '"')
    ret = ret.replace("'", '"')

    # Convert to upper case with unicode normalization
    ret = unicode_upper(ret)

    # Normalize quotes again after uppercasing (in case NFKD changed them)
    ret = ret.replace('«', '"')
    ret = ret.replace('»', '"')
    ret = ret.replace("'", '"')

    # Remove all punctuation except:
    # - " (double quotes are kept as canonical quote marker)
    # - ? and ! (only at the end, 1-4 chars)
    # - parentheses (significant characters)

    # First, collect trailing ? and ! (1-4 chars) from the end
    trailing_punct = ""
    i = len(ret) - 1
    while i >= 0 and len(trailing_punct) < 4 and ret[i] in '?!':
        trailing_punct = ret[i] + trailing_punct
        i -= 1

    # Remove all punctuation except parentheses and double quotes from the string
    punctuation_to_remove = set(str_lib.punctuation) - {'(', ')', '"'}

    result = []
    for char in ret:
        if char in punctuation_to_remove:
            continue
        result.append(char)

    ret = ''.join(result)

    # Strip multiple spaces again after punctuation removal
    while '  ' in ret:
        ret = ret.replace('  ', ' ')

    # Strip again after removing punctuation
    ret = ret.strip()

    # Add back trailing punctuation if any
    if trailing_punct:
        ret = ret + trailing_punct

    return ret

File: app_go/util/strings.go
// Package util provides string manipulation utilities.
package util

import (
	"crypto/md5"
	"fmt"
	"strings"
	"unicode"

	"golang.org/x/text/unicode/norm"
)

// REPLACEMENT_MAP maps special characters to their normalized string equivalents.
var REPLACEMENT_MAP = map[rune]string{
	// Cyrillic
	'Ё': "Е",
	'Й': "И",
	'Ъ': "Ь",
	// German umlauts (uppercase)
	'Ä': "AE",
	'Ö': "OE",
	'Ü': "UE",
	'ß': "SS",
	// Other common replacements (uppercase)
	'À': "A", 'Á': "A", 'Â': "A", 'Ã': "A", 'Å': "A",
	'È': "E", 'É': "E", 'Ê': "E", 'Ë': "E",
	'Ì': "I", 'Í': "I", 'Î': "I", 'Ï': "I",
	'Ñ': "N",
	'Ò': "O", 'Ó': "O", 'Ô': "O", 'Õ': "O",
	'Ù': "U", 'Ú': "U", 'Û': "U",
	'Ý': "Y",
	'Ç': "C",
	'Þ': "TH",
	'Đ': "D",
	'Ł': "L",
	'Ń': "N",
	'Ǿ': "O",
	'Ŕ': "R",
	'Ś': "S",
	'Ź': "Z",
	'Ż': "Z",
}

// UnicodeUpper performs uppercase conversion with NFKD normalization.
func UnicodeUpper(s string) string {
	// NFKD normalize
	transformed := norm.NFKD.String(s)

	// Remove combining characters (category Mn = Nonspacing_Mark)
	var buf strings.Builder
	for _, r := range transformed {
		if unicode.Is(unicode.Mn, r) {
			continue
		}
		buf.WriteRune(r)
	}
	ret := buf.String()

	// Upper case
	ret = strings.ToUpper(ret)

	// Apply replacements (handle multi-char replacements like Ä->AE)
	var result strings.Builder
	for _, r := range ret {
		if repl, ok := REPLACEMENT_MAP[r]; ok {
			result.WriteString(repl)
		} else {
			result.WriteRune(r)
		}
	}
	return result.String()
}

// StrNormalize normalizes a string for ID generation and comparison.
func StrNormalize(s string) string {
	if s == "" {
		return ""
	}

	// Strip leading/trailing whitespace
	ret := strings.TrimSpace(s)

	// Replace multiple spaces with single space
	for strings.Contains(ret, "  ") {
		ret = strings.ReplaceAll(ret, "  ", " ")
	}

	// Normalize all quote types to double quotes
	ret = strings.ReplaceAll(ret, "«", "\"")
	ret = strings.ReplaceAll(ret, "»", "\"")
	ret = strings.ReplaceAll(ret, "'", "\"")

	// Upper case with unicode normalization
	ret = UnicodeUpper(ret)

	// Normalize quotes again after uppercasing
	ret = strings.ReplaceAll(ret, "«", "\"")
	ret = strings.ReplaceAll(ret, "»", "\"")
	ret = strings.ReplaceAll(ret, "'", "\"")

	// Collect trailing ? and ! (1-4 chars)
	trailingPunct := ""
	i := len(ret) - 1
	for i >= 0 && len(trailingPunct) < 4 {
		ch := ret[i]
		if ch == '?' || ch == '!' {
			trailingPunct = string(ch) + trailingPunct
			i--
		} else {
			break
		}
	}

	// Remove punctuation except parentheses and double quotes
	var result strings.Builder
	for _, ch := range ret {
		switch ch {
		case '(', ')', '"':
			result.WriteRune(ch)
		default:
			if isPunctuation(ch) {
				continue
			}
			result.WriteRune(ch)
		}
	}
	ret = result.String()

	// Strip multiple spaces after punctuation removal
	for strings.Contains(ret, "  ") {
		ret = strings.ReplaceAll(ret, "  ", " ")
	}

	// Strip again
	ret = strings.TrimSpace(ret)

	// Add back trailing punctuation
	if trailingPunct != "" {
		ret = ret + trailingPunct
	}

	return ret
}

// isPunctuation checks if a rune is a punctuation character.
func isPunctuation(r rune) bool {
	punctuation := "`-~=!@#$%^&*_+[]{}\\|;:',.<>/"
	for _, p := range punctuation {
		if r == p {
			return true
		}
	}
	return false
}

// MakeID generates an MD5 hex ID from a normalized string.
func MakeID(name string, nameAsIs bool) string {
	nameStr := "--- unknown ---"
	if name != "" {
		nameStr = name
	}
	normName := nameStr
	if !nameAsIs {
		normName = StrNormalize(nameStr)
	}
	hash := md5.Sum([]byte(normName))
	return fmt.Sprintf("%x", hash[:])
}

// ID2Path converts an ID to a directory path: "1a2b3c..." -> "1a/2b/1a2b3c..."
func ID2Path(id string) string {
	if len(id) < 4 {
		return id
	}
	first := id[:2]
	second := id[2:4]
	return first + "/" + second + "/" + id
}

// ID2PathOnly returns the directory part of ID2Path: "1a2b3c..." -> "1a/2b"
func ID2PathOnly(id string) string {
	if len(id) < 4 {
		return id
	}
	first := id[:2]
	second := id[2:4]
	return first + "/" + second
}

// URLStr URL-encodes a string with custom character replacements.
func URLStr(s string) string {
	if s == "" {
		return ""
	}
	transl := map[byte]string{
		'"':  "%22",
		'\'': "%27",
	}
	var result strings.Builder
	for i := 0; i < len(s); i++ {
		ch := s[i]
		if replacement, ok := transl[ch]; ok {
			result.WriteString(replacement)
		} else {
			result.WriteByte(ch)
		}
	}
	// Now percent-encode the result
	return strings.ReplaceAll(result.String(), " ", "%20")
}

// StripQuotes safely removes symmetric quotes from a string.
func StripQuotes(s string) string {
	if s == "" {
		return ""
	}
	s = strings.TrimSpace(s)
	if len(s) >= 2 {
		if (s[0] == '"' && s[len(s)-1] == '"') || (s[0] == '\'' && s[len(s)-1] == '\'') {
			return s[1 : len(s)-1]
		}
	}
	return s
}